File: src/commands/boxCommands.ts
// src/commands/boxCommands.ts
import { core } from "@tauri-apps/api";
import type { Box } from "../types";

export const boxCommands = {
    async getAllBoxes(): Promise<Box[]> {
        return await core.invoke("get_all_boxes");
    },

    async getBox(boxId: string): Promise<Box> {
        return await core.invoke("get_box", { boxId });
    },

    async createBox(
        name: string,
        description: string | null,
        tags: string[],
        devMode: boolean
    ): Promise<string> {
        return await core.invoke("create_box", {
            name,
            description,
            tags,
            devMode
        });
    },

    async updateBox(
        boxId: string,
        name: string | null,
        description: string | null | undefined,
        tags: string[] | null,
        devMode: boolean | null
    ): Promise<void> {
        return await core.invoke("update_box", {
            boxId,
            name,
            description,
            tags,
            devMode
        });
    },

    async deleteBox(boxId: string): Promise<void> {
        return await core.invoke("delete_box", { boxId });
    },

    async deleteSelectedBoxes(boxIds: string[]): Promise<string[]> {
        return await core.invoke("delete_selected_boxes", { boxIds });
    },

    async searchBoxesGlobal(query: string, tags: string[]): Promise<Box[]> {
        return await core.invoke("search_boxes_global", { query, tags });
    },
};

File: src-tauri/src/commands/box_commands.rs
// src/commands/box_commands.rs

use crate::models::Box;
use crate::states::AppState;
use std::sync::Arc;
use tauri::State;

#[tauri::command]
pub async fn get_all_boxes(app_state: State<'_, Arc<AppState>>) -> Result<Vec<Box>, String> {
    app_state.get_all_boxes().await.map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_box(box_id: String, app_state: State<'_, Arc<AppState>>) -> Result<Box, String> {
    app_state.get_box(&box_id).await.map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_box(
    name: String,
    description: Option<String>,
    tags: Vec<String>,
    dev_mode: bool, 
    app_state: State<'_, Arc<AppState>>,
) -> Result<String, String> {
    app_state
        .create_box(name, description, tags, dev_mode)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn update_box(
    box_id: String,
    name: Option<String>,
    description: Option<Option<String>>,
    tags: Option<Vec<String>>,
    dev_mode: Option<bool>,
    app_state: State<'_, Arc<AppState>>,
) -> Result<(), String> {
    app_state
        .update_box(&box_id, name, description, tags, dev_mode)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn delete_box(
    box_id: String,
    app_state: State<'_, Arc<AppState>>,
) -> Result<(), String> {
    app_state
        .delete_box(&box_id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn delete_selected_boxes(
    box_ids: Vec<String>,
    app_state: State<'_, Arc<AppState>>,
) -> Result<Vec<String>, String> {
    app_state
        .delete_selected_boxes(box_ids)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn search_boxes_global(
    query: String,
    tags: Vec<String>,
    app_state: State<'_, Arc<AppState>>,
) -> Result<Vec<Box>, String> {
    app_state
        .search_boxes_global(query, tags)
        .await
        .map_err(|e| e.to_string())
}
